File: problems/Design Browser History/browserHistory.cs
public class BrowserHistory {

    private Stack<string> _backStack = new Stack<string>();
    private Stack<string> _forwardStack = new Stack<string>();
    private string _currUrl;

    public BrowserHistory(string homepage) {
        _currUrl = homepage;
    }

    public void Visit(string url) {
        _backStack.Push(_currUrl);
        _forwardStack.Clear();
        _currUrl = url;
    }

    public string Back(int steps) {
        if (0 < _backStack.Count) {
            _forwardStack.Push(_currUrl);
        }

        while (1 < steps-- && 1 < _backStack.Count) {
            _forwardStack.Push(_backStack.Pop());
        }

        if (0 < _backStack.Count) {
            _currUrl = _backStack.Pop();
        }

        return _currUrl;
    }

    public string Forward(int steps) {
        if (0 < _forwardStack.Count) {
            _backStack.Push(_currUrl);
        }

        while (1 < steps-- && 1 < _forwardStack.Count) {
            _backStack.Push(_forwardStack.Pop());
        }

        if (0 < _forwardStack.Count) {
            _currUrl = _forwardStack.Pop();
        }

        return _currUrl;
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.Visit(url);
 * string param_2 = obj.Back(steps);
 * string param_3 = obj.Forward(steps);
 */

File: problems/Design Browser History/browserHistory.java
class BrowserHistory {

    public BrowserHistory(String homepage) {
        backStack = new Stack<>();
        forwardStack = new Stack<>();
        currUrl = homepage;
    }

    public void visit(String url) {
        backStack.push(currUrl);
        forwardStack.clear();
        currUrl = url;
    }

    public String back(int steps) {
        if (!backStack.isEmpty()) {
            forwardStack.push(currUrl);
        }

        while (1 < steps-- && 1 < backStack.size()) {
            forwardStack.push(backStack.pop());
        }

        if (!backStack.isEmpty()) {
            currUrl = backStack.pop();
        }

        return currUrl;
    }

    public String forward(int steps) {
        if (!forwardStack.isEmpty()) {
            backStack.push(currUrl);
        }

        while (1 < steps-- && 1 < forwardStack.size()) {
            backStack.push(forwardStack.pop());
        }

        if (!forwardStack.isEmpty()) {
            currUrl = forwardStack.pop();
        }

        return currUrl;
    }

    private Stack<String> backStack;
    private Stack<String> forwardStack;
    private String currUrl;
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */
